File: server/proto/mqtt.go
package proto

type MqttCommand struct {
	Name    string `json:"name"`
	Topic   string `json:"topic"`
	Payload string `json:"payload"`
}

type GetMqttConfigRsp struct {
	Enabled  bool          `json:"enabled"`
	Broker   string        `json:"broker"`
	Port     int           `json:"port"`
	Tls      bool          `json:"tls"`
	Username string        `json:"username"`
	Topic    string        `json:"topic"`
	Commands []MqttCommand `json:"commands"`
	// The password itself is never returned.
	HasPassword bool `json:"hasPassword"`

	HaEnabled         bool   `json:"haEnabled"`
	HaDiscoveryPrefix string `json:"haDiscoveryPrefix"`
	HaNodeId          string `json:"haNodeId"`
	HaDeviceName      string `json:"haDeviceName"`
}

type SetMqttConfigReq struct {
	Enabled  *bool  `json:"enabled"`
	Broker   string `json:"broker"`
	Port     int    `json:"port"`
	Tls      *bool  `json:"tls"`
	Username string `json:"username"`
	Topic    string `json:"topic"`
	// Omitted leaves the stored password unchanged; an empty string clears it.
	Password *string       `json:"password"`
	Commands []MqttCommand `json:"commands"`

	// Pointers so an omitted field leaves the stored value untouched.
	HaEnabled         *bool   `json:"haEnabled"`
	HaDiscoveryPrefix *string `json:"haDiscoveryPrefix"`
	HaNodeId          *string `json:"haNodeId"`
	HaDeviceName      *string `json:"haDeviceName"`
}

type PublishMqttReq struct {
	Name string `json:"name" form:"name" validate:"required"`
}

File: web/src/api/mqtt.ts
import { http } from '@/lib/http.ts';

export type MqttCommand = {
  name: string;
  topic: string;
  payload: string;
};

export type MqttConfig = {
  enabled: boolean;
  broker: string;
  port: number;
  tls: boolean;
  username: string;
  topic: string;
  commands: MqttCommand[];
  hasPassword: boolean;
};

// The stored password is never sent to the client; omitting it here keeps it
// unchanged on the device, while an empty string clears it.
export type MqttConfigUpdate = Omit<MqttConfig, 'hasPassword'> & {
  password?: string;
};

export function getMqttConfig() {
  return http.get('/api/mqtt/config');
}

export function setMqttConfig(config: MqttConfigUpdate) {
  return http.post('/api/mqtt/config', config);
}

export function publishMqttCommand(name: string) {
  return http.post('/api/mqtt/publish', { name });
}
